lnwallet/channel.go

```diff
@@ -2334,133 +2343,280 @@ func NewBreachRetribution(chanState *channeldb.OpenChannel, stateNum uint64,
+	)
+	if err != nil {
+		return emptyRetribution, err
+	}
+
+	return HtlcRetribution{
+		SignDesc: input.SignDescriptor{
+			KeyDesc: chanState.LocalChanCfg.
+				RevocationBasePoint,
+			DoubleTweak:   commitmentSecret,
+			WitnessScript: htlcWitnessScript,
+			Output: &wire.TxOut{
+				PkScript: htlcPkScript,
+				Value:    int64(htlc.Amt.ToSatoshis()),
+			},
+			HashType: txscript.SigHashAll,
+		},
+		OutPoint: wire.OutPoint{
+			Hash:  commitHash,
+			Index: uint32(htlc.OutputIndex),
+		},
+		SecondLevelWitnessScript: secondLevelScript.WitnessScript,
+		IsIncoming:               htlc.Incoming,
+	}, nil
+}
+
+// createBreachRetribution creates a partially initiated BreachRetribution
+// using a RevocationLog. Returns the constructed retribution, our amount,
+// their amount, and a possible non-nil error.
+func createBreachRetribution(revokedLog *channeldb.RevocationLog,
+	spendTx *wire.MsgTx, chanState *channeldb.OpenChannel,
+	keyRing *CommitmentKeyRing, commitmentSecret *btcec.PrivateKey,
+	leaseExpiry uint32) (*BreachRetribution, int64, int64, error) {
+
+	commitHash := revokedLog.CommitTxHash
+
+	// Construct the outpoints.
+	ourOutpoint := wire.OutPoint{
+		Hash:  commitHash,
+		Index: uint32(revokedLog.OurOutputIndex),
+	}
+	theirOutpoint := wire.OutPoint{
+		Hash:  commitHash,
+		Index: uint32(revokedLog.TheirOutputIndex),
+	}
+
+	// Create the htlc retributions.
+	htlcRetributions := make([]HtlcRetribution, len(revokedLog.HTLCEntries))
+	for i, htlc := range revokedLog.HTLCEntries {
+		hr, err := createHtlcRetribution(
+			chanState, keyRing, commitHash,
+			commitmentSecret, leaseExpiry, htlc,
 		)
 		if err != nil {
-			return nil, err
+			return nil, 0, 0, err
 		}
+		htlcRetributions[i] = hr
+	}
 
-		// If this is an incoming HTLC, then this means that they were
-		// the sender of the HTLC (relative to us). So we'll
-		// re-generate the sender HTLC script. Otherwise, is this was
-		// an outgoing HTLC that we sent, then from the PoV of the
-		// remote commitment state, they're the receiver of this HTLC.
-		htlcPkScript, htlcWitnessScript, err := genHtlcScript(
-			chanState.ChanType, htlc.Incoming, false,
-			htlc.RefundTimeout, htlc.RHash, keyRing,
+	// Sanity check that OurOutputIndex is within range.
+	if int(revokedLog.OurOutputIndex) >= len(spendTx.TxOut) {
+		return nil, 0, 0, fmt.Errorf("%w: ours=%v, len(TxOut)=%v",
+			ErrOutputIndexOutOfRange,
+			revokedLog.OurOutputIndex, len(spendTx.TxOut),
 		)
-		if err != nil {
-			return nil, err
+	}
+
+	// Sanity check that TheirOutputIndex is within range.
+	if int(revokedLog.TheirOutputIndex) >= len(spendTx.TxOut) {
+		return nil, 0, 0, fmt.Errorf("%w: theirs=%v, len(TxOut)=%v",
+			ErrOutputIndexOutOfRange,
+			revokedLog.TheirOutputIndex, len(spendTx.TxOut),
+		)
+	}
+
+	// Read the amounts from the breach transaction.
+	//
+	// NOTE: ourAmt here includes commit fee and anchor amount(if enabled).
+	ourAmt := spendTx.TxOut[revokedLog.OurOutputIndex].Value
```

Comment:
If one of the outputs doesn't exist, I think the index will be zero here and then `ourAmt`/`theirAmt` would have the wrong value. Then a sign descriptor would get incorrectly created in `NewBreachRetribution`

Reply:
Good point! Maybe use a max uint32 to init these fields?

Reply:
Also curious about how we are handling this now, will check.

Reply:
So we would check whether the local or remote `OutputSignDesc` against nil to handle this.